File: kin-sdk/src/main/java/kin/sdk/KinAccountImpl.java
package kin.sdk;


import java.math.BigDecimal;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import kin.base.KeyPair;
import kin.sdk.exception.AccountDeletedException;
import kin.sdk.exception.CryptoException;
import kin.sdk.exception.OperationFailedException;
import kin.utils.MainHandler;


final class KinAccountImpl extends AbstractKinAccount {

    private final KeyPair account;
    private final BackupRestore backupRestore;
    private final TransactionSender transactionSender;
    private final AccountInfoRetriever accountInfoRetriever;
    private final BlockchainEvents blockchainEvents;
    private boolean isDeleted = false;

    KinAccountImpl(KeyPair account, BackupRestore backupRestore, TransactionSender transactionSender,
        AccountInfoRetriever accountInfoRetriever, BlockchainEventsCreator blockchainEventsCreator, MainHandler mainHandler) {
        super(mainHandler);

        this.account = account;
        this.backupRestore = backupRestore;
        this.transactionSender = transactionSender;
        this.accountInfoRetriever = accountInfoRetriever;
        this.blockchainEvents = blockchainEventsCreator.create(account.getAccountId());
    }

    @Override
    public String getPublicAddress() {
        if (!isDeleted) {
            return account.getAccountId();
        }
        return null;
    }

    @Override
    public Transaction buildTransactionSync(@Nonnull String publicAddress, @Nonnull BigDecimal amount,
                                            int fee) throws OperationFailedException {
        checkValidAccount();
        return transactionSender.buildTransaction(account, publicAddress, amount, fee);
    }

    @Override
    public Transaction buildTransactionSync(@Nonnull String publicAddress, @Nonnull BigDecimal amount,
                                            int fee, @Nullable String memo) throws OperationFailedException {
        checkValidAccount();
        return transactionSender.buildTransaction(account, publicAddress, amount, fee, memo);
    }

    @Nonnull
    @Override
    public TransactionId sendTransactionSync(Transaction transaction) throws OperationFailedException {
        checkValidAccount();
        return transactionSender.sendTransaction(transaction);
    }

    @Nonnull
    @Override
    public TransactionId sendWhitelistTransactionSync(String whitelist) throws OperationFailedException {
        checkValidAccount();
        return transactionSender.sendWhitelistTransaction(whitelist);
    }

    @Nonnull
    @Override
    public Balance getBalanceSync() throws OperationFailedException {
        checkValidAccount();
        return accountInfoRetriever.getBalance(account.getAccountId());
    }

    @Override
    public AccountStatus getStatusSync() throws OperationFailedException {
        checkValidAccount();
        return accountInfoRetriever.getStatus(account.getAccountId());
    }

    @Override
    public ListenerRegistration addBalanceListener(@Nonnull EventListener<Balance> listener) {
        return blockchainEvents.addBalanceListener(listener);
    }

    @Override
    public ListenerRegistration addPaymentListener(@Nonnull EventListener<PaymentInfo> listener) {
        return blockchainEvents.addPaymentListener(listener);
    }

    @Override
    public ListenerRegistration addAccountCreationListener(EventListener<Void> listener) {
        return blockchainEvents.addAccountCreationListener(listener);
    }

    @Override
    public String export(@Nonnull String passphrase) throws CryptoException {
        return backupRestore.exportWallet(account, passphrase);
    }

    void markAsDeleted() {
        isDeleted = true;
    }

    private void checkValidAccount() throws AccountDeletedException {
        if (isDeleted) {
            throw new AccountDeletedException();
        }
    }

}

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/internal/KinAccountImpl.kt
package kin.sdk.internal

import kin.base.KeyPair
import kin.sdk.KinAccount
import kin.sdk.exception.AccountDeletedException
import kin.sdk.exception.CryptoException
import kin.sdk.exception.OperationFailedException
import kin.sdk.internal.services.AccountInfoRetriever
import kin.sdk.internal.services.BlockchainEvents
import kin.sdk.internal.services.TransactionSender
import kin.sdk.internal.services.helpers.EventListener
import kin.sdk.internal.services.helpers.ListenerRegistration
import kin.sdk.internal.utils.BackupRestore
import kin.sdk.internal.utils.BlockchainEventsCreator
import kin.sdk.models.Balance
import kin.sdk.models.PaymentInfo
import kin.sdk.models.Transaction
import kin.sdk.models.TransactionId
import kin.utils.Request
import java.math.BigDecimal
import java.util.concurrent.Callable

internal class KinAccountImpl(
    private val account: KeyPair,
    private val backupRestore: BackupRestore,
    private val transactionSender: TransactionSender,
    private val accountInfoRetriever: AccountInfoRetriever,
    blockchainEventsCreator: BlockchainEventsCreator
) : KinAccount {
    private val blockchainEvents: BlockchainEvents by lazy { blockchainEventsCreator.create(account.accountId) }
    private var isDeleted = false

    override val publicAddress: String?
        get() = if (!isDeleted) {
            account.accountId
        } else null

    override val balanceSync: Balance
        @Throws(OperationFailedException::class)
        get() = checkValidAccount()
            .let { accountInfoRetriever.getBalance(account.accountId) }

    override val statusSync: Int
        @Throws(OperationFailedException::class)
        get() = checkValidAccount()
            .let { accountInfoRetriever.getStatus(account.accountId) }

    override val balance: Request<Balance>
        get() = Request(Callable<Balance> { balanceSync })

    override val status: Request<Int>
        get() = Request(Callable<Int> { statusSync })

    @Throws(OperationFailedException::class)
    override fun buildTransactionSync(
        publicAddress: String,
        amount: BigDecimal,
        fee: Int
    ): Transaction =
        checkValidAccount()
            .let {
                transactionSender.buildTransaction(account, publicAddress, amount, fee)
            }

    @Throws(OperationFailedException::class)
    override fun buildTransactionSync(publicAddress: String, amount: BigDecimal,
                                      fee: Int, memo: String?): Transaction =
        checkValidAccount()
            .let {
                transactionSender.buildTransaction(account, publicAddress, amount, fee, memo)
            }

    @Throws(OperationFailedException::class)
    override fun sendTransactionSync(transaction: Transaction): TransactionId =
        checkValidAccount()
            .let {
                transactionSender.sendTransaction(transaction)
            }

    @Throws(OperationFailedException::class)
    override fun sendWhitelistTransactionSync(whitelist: String): TransactionId =
        checkValidAccount()
            .let { transactionSender.sendWhitelistTransaction(whitelist) }

    override fun addBalanceListener(listener: EventListener<Balance>): ListenerRegistration =
        blockchainEvents.addBalanceListener(listener)

    override fun addPaymentListener(listener: EventListener<PaymentInfo>): ListenerRegistration =
        blockchainEvents.addPaymentListener(listener)

    override fun addAccountCreationListener(listener: EventListener<Void?>): ListenerRegistration =
        blockchainEvents.addAccountCreationListener(listener)

    @Throws(CryptoException::class)
    override fun export(passphrase: String): String =
        backupRestore.exportAccount(account, passphrase).jsonString

    fun markAsDeleted() {
        isDeleted = true
    }

    @Throws(AccountDeletedException::class)
    private fun checkValidAccount() {
        if (isDeleted) {
            throw AccountDeletedException()
        }
    }

    override fun buildTransaction(
        publicAddress: String,
        amount: BigDecimal,
        fee: Int
    ): Request<Transaction> = Request { buildTransactionSync(publicAddress, amount, fee) }

    override fun buildTransaction(
        publicAddress: String,
        amount: BigDecimal,
        fee: Int,
        memo: String?
    ): Request<Transaction> = Request { buildTransactionSync(publicAddress, amount, fee, memo) }

    override fun sendTransaction(transaction: Transaction): Request<TransactionId> =
        Request { sendTransactionSync(transaction) }

    override fun sendWhitelistTransaction(whitelist: String): Request<TransactionId> =
        Request { sendWhitelistTransactionSync(whitelist) }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val account = other as KinAccount?
        return if (publicAddress == null || account!!.publicAddress == null) {
            false
        } else publicAddress == account.publicAddress
    }

    override fun hashCode(): Int {
        var result = account.hashCode()
        result = 31 * result + backupRestore.hashCode()
        result = 31 * result + transactionSender.hashCode()
        result = 31 * result + accountInfoRetriever.hashCode()
        result = 31 * result + blockchainEvents.hashCode()
        result = 31 * result + isDeleted.hashCode()
        return result
    }
}
